guidelines/terms/22/target-offset.html

```diff
@@ -1,6 +1,11 @@
 <dt class="new"><dfn data-lt="target offsets">target offset</dfn></dt>
 <dd class="new">
 	<p class="change">New</p>	
-   <p>the distance measured from the farthest point of a target (A), to the nearest point of the second target (B). The offset includes the target and spacing around the target.</p>
+   <p>size of a target (A) plus the spacing to a second target (B).</p>
    <p class="note">The offset from A to B may be different than the offset from B to A, if the sizes of these targets differ.</p>
+   <div class="note">
+      <p>For horizontally aligned targets, target offset is the width of target A plus the horizontal space to the closest edge of target B. For vertically aligned targets, target offset is the the height of target A plus the vertical space to the closest edge of target B. For targets that are neither horizontally nor vertically aligned, the target offset is the longest measure from any point of A to the nearest point of B.</p>
```

Comment:
"the closest edge of target B" feels imprecisely worded for targets that aren't rectangular. How should this be interpreted if B is a circle?

Reply:
(it's not as simple as just "the nearer of the left and right edges of target B's bounding box" because that would do the wrong thing in the case of targets which are both aligned and overlapping)

Reply:
If B is a circle, I think the closest edge would be the closest point.

In fact, I guess we could just say the closest point and it would work for all cases?

If a target is overlapping, target offset is not needed and you check for 24 by 24px available in each target.

Reply:
"Closest point" is less ambiguous than "closest edge" but still gets us back to the same ambiguity around what to interpret it relative to that we've discussed ad nauseam in #2695. What point in (A) do you measure from (particularly if A is non-rectangular)? What is "closest" relative to?

For example, here are two side-by-side circular targets which are "horizontally aligned" per this note, but with the left one' center slightly higher than the right one's:

![image](https://user-images.githubusercontent.com/376284/202558587-beafc4cb-da46-4660-a01a-91b2b45c780e.png)

In this scenario, what is the "horizontal space to the closest point of target B"? Should a reader interpret this as "for every possible tuple of one point in A and one in B, find the difference in those two points' X values, and take the minimum of all those differences"? Or "for every possible tuple of one point in A and one in B, find the distance between those points; take the tuple with the minimum distance, and then use the difference in X values of those two points"? Or maybe one of the first two definitions but restricted to "rightmost point in A" and "leftmost point in B", rather than "any point"? I can imagine reasonable readers interpreting it in any of these ways, and they result in 3 different values in this example.

Reply:
> If a target is overlapping, target offset is not needed and you check for 24 by 24px available in each target.

I think this is a reasonable intent, but it doesn't match the writing in the current proposal. As currently written in this proposal, nothing prevents target offset from applying to overlapping targets. Here is an example with a 20x24px target B overlaid on top of a 20x40px target A such that A underhangs B on both the left and right sides.

![rendering of the 20x24px over 20x40px scenario described above](https://user-images.githubusercontent.com/376284/202562254-7df328f7-6c3b-426e-ad2d-88076ade8cad.png)

Per the current proposal, these are "horizontally aligned", and since B is 24px wide, it passes the Spacing exception regardless of how its spacing to A is calculated. I think as currently written it's ambiguous whether A passes the Spacing exception - a reasonable reader might conclude that A is "40px wide" and so passes, but I think the intended behavior is to fail such a case.

Reply:
With this two circles example, calculating target offset is not too bad:  diameter plus length of any a horizontal line between -- which is not as intuitive as diameter plus measure of closest distance.  But measuring 'size' of oddly shaped targets (needed when target are not aligned) seems fickle.

Reply:
> 



> With this two circles example, calculating target offset is not too bad: diameter plus length of any a horizontal line between -- which is a larger value than, and not as intuitive as, diameter plus spacing.

"any a horizontal line" is not specific enough; there are many different-length horizontal lines that you could draw in this example. You get different values for offset if you choose to start from "the rightmost point of A" vs "the point of A closest to some point of B" vs "the point of A which minimizes the horizontal distance to the closest point of B". The normative wording of the spec needs to define the term unambiguously if it's going to be the basis of an SC's pass/fail conditions.

> With your 'overlapping' rectangles, I think you would have to treat that as three targets, no?

The definition of "target" in the current draft is "region of the display that will accept a pointer action, such as the interactive area of a user interface component". This definition does not require that it be a *contiguous* region, so I don't think you would *have to* treat that as three targets, though I think you could if you wanted to. Practically speaking, the image I gave is from a bit of HTML that involves 2 buttons, not 3 buttons, so I think it's likely that at least some readers would interpret it as 2 targets in practice. That said, I do think it's likely that different readers will interpret that differently; it might be nice for the target definition to include a note clarifying.
 
> But consider if A is 2 pixels taller, so then A and B are no longer horizontally aligned. Now what is the 'size' of A?
> 
> These examples did make me realize that measuring 'size' of oddly shaped targets (needed when target are not aligned) seems fickle.
> 
> If ellipses instead of circles, using 'diameter' is no good. Bounding box then?

Yes, I agree that the terms "width" and "size" are problematic for oddly shaped targets. I think that's actually problematic for the primary clause of the 2.5.8 too. Is a 40x40px square button with a different 20x20px button overlaid on top of it "at least 24 by 24 CSS pixels"?

![40x40px square target A with 20x20px square target B overlaid in the middle](https://user-images.githubusercontent.com/376284/202600844-745cbdf9-d8b8-42d8-8f61-c6150725e661.png)

I think we obviously want target A in this example **not** to be considered to meet the 24x24 requirement, but I think different readers of the current wording of the SC will disagree about whether such a button meets the requirement.

Reply:
Yes, presuming overlap of A to be thin, I agree it must not be considered to meet the Spacing exception.

With the present wording -- "size plus spacing" -- I think were are okay.  One has to allow a bad-faith assertion about the size of A (e.g., length of an outside side) rather than inside thickness.  Size measurement taken in good faith can be  covered by Understanding.

Several SC have loopholes allowing for terrible design or malicious compliance, but that's okay if balanced against phrasing which is simpler to understand.

With your circles example, the measure for "horizonal width" is obviously the diameter.  If B is up high enough, yes I can imagine someone naively extending the horizontal diameter line from A to the top (or near top) of B, and that would be bad.

Right now, (as I understand it) the current sentence under consideration is:

> For horizontally aligned targets, target offset is the width of target A plus the horizontal space to the closest point of target B.

How about:

> For horizontally aligned targets, target offset is the width of target A plus the smallest horizontal measurement between A and the closest point of target B.

Ping to @WilcoFiers

Reply:
I've broken out the conversation about "size" and "24 x 24" being ambiguous to a separate issue (#2803), since they're not really about these changes.

That said, I think the term "width" here has the same problems that #2803 describes, where reasonable interpretations of it behave poorly for overlapped targets. I'm struggling to come up with a reasonably understandable equivalent that is not ambiguous. The closest I've come up with is

> For horizontally aligned targets, target offset is the length of the longest horizontal line with one endpoint touching a boundary of A, the other endpoint touching the bounding rectangle of B, and which never intersects any non-A target except at the endpoints.

...but this is pretty rough to understand in plain language and feels really questionable in terms of how far off it is from the original intent of the SC.